File: src/app/api/langchain_bot/langchain.js
const { formatDocumentsAsString } = require("langchain/util/document");
const { createChatbotComponents } = require("./utils");
const { RunnableSequence, RunnablePassthrough } = require("@langchain/core/runnables");
const { StringOutputParser } = require("@langchain/core/output_parsers");

const langchainInvoke = async (msg) => {
    const { retriever, prompt, model } = await createChatbotComponents();

    const chain = RunnableSequence.from([
        {
            context: retriever.pipe(formatDocumentsAsString),
            question: new RunnablePassthrough(),
        },
        prompt,
        model,
        new StringOutputParser(),
    ]);

    const res = await chain.invoke(msg)
    return res;
}

exports.langchainInvoke = langchainInvoke

File: lambda/python-monocle-s3-langchain/coffee-chatbot-python/chatbot_lib/langchain.py
# from langchain.utils import format_documents_as_string
from langchain_core.runnables import RunnableSequence, RunnablePassthrough
from langchain_core.output_parsers import StrOutputParser
from .utils import create_chatbot_components, format_docs

def langchain_invoke(msg):
    retriever, prompt, model = create_chatbot_components()
    rag_chain = (
        {"context": retriever | format_docs, "question": RunnablePassthrough()}
        | prompt
        | model
        | StrOutputParser()
    )
    val = rag_chain.invoke(msg)
    return val
